### PR TITLE
Failed to handle large bulk reply

### PR DESCRIPTION
**Pull request by bubriks**
[#75](https://github.com/Wappsto/wappsto-python/issues/75)

### CODE REVIEW

File: test/wappsto_test.py [hunk 1]
```diff
@@ -52,22 +53,36 @@ def get_object(self, object_name):
...
-def create_response(self, verb, trace_id, bulk, id, url, data):
+def create_response(self, verb, trace_id, bulk, id, url, data, split_message):
     trace = ''
 
     if verb == "DELETE" or verb == "PUT" or verb == "GET":
         if trace_id is not None:
             trace = {"trace": str(trace_id)}
 
-        message = {"jsonrpc": "2.0", "id": "1", "params": {"url": str(url), "meta": trace, "data": {"meta": {"id": id}, "data": data}}, "method": verb}
+        message = {"jsonrpc": "2.00", "id": "1", "params": {"url": str(url), "meta": trace, "data": {"meta": {"id": id}, "data": data}}, "method": verb}
```
Comment (AndreasBomholtz) [inline test/wappsto_test.py]: Why did you change this?

File: wappsto/connection/communication.py [hunk 2]
```diff
@@ -618,19 +619,24 @@ def receive_data(self):
...
+                message = ''.join(total_decoded)
+                try:
+                    decoded = json.loads(message)
+                except JSONDecodeError:
+                    if len(decoded_data) < MESSAGE_SIZE:
+                        total_decoded = []
+                        if data == b'':
+                            self.reconnect()
+                        else:
+                            error = "Value error: {}".format(message)
```
Comment (AndreasBomholtz) [inline wappsto/connection/communication.py]: It is not a value error. It is json decoding error.

File: wappsto/connection/communication.py [hunk 3]
```diff
@@ -618,19 +619,24 @@ def receive_data(self):
...
-            except JSONDecodeError:
-                if data == b'':
-                    self.reconnect()
+                message = ''.join(total_decoded)
+                try:
+                    decoded = json.loads(message)
+                except JSONDecodeError:
+                    if len(decoded_data) < MESSAGE_SIZE:
+                        total_decoded = []
+                        if data == b'':
```
Comment (AndreasBomholtz) [inline wappsto/connection/communication.py]: Why not check for connection error right after it has been read?

File: test/wappsto_test.py [hunk 4]
```diff
@@ -52,22 +53,36 @@ def get_object(self, object_name):
     return actual_object
 
 
-def create_response(self, verb, trace_id, bulk, id, url, data):
+def create_response(self, verb, trace_id, bulk, id, url, data, split_message):
```
Comment (AndreasBomholtz) [inline test/wappsto_test.py]: Maybe rename to send response?

File: wappsto/connection/communication.py [hunk 5]
```diff
@@ -618,19 +619,24 @@ def receive_data(self):
         decoded = None
         while True:
             if self.connected:
-                data = self.my_socket.recv(2000)
+                data = self.my_socket.recv(MESSAGE_SIZE)
                 decoded_data = data.decode('utf-8')
```
Comment (AndreasBomholtz) [inline wappsto/connection/communication.py]: You should add ignore to the decode, to make sure that there is no exceptions thrown.

File: wappsto/connection/communication.py [hunk 6]
```diff
@@ -618,19 +619,24 @@ def receive_data(self):
...
                 total_decoded.append(decoded_data)
```
Comment (AndreasBomholtz) [inline wappsto/connection/communication.py]: Maybe there also should be max size for the combined size. Like 1Mb...

File: wappsto/connection/communication.py [hunk 7]
```diff
@@ -623,23 +626,31 @@ def receive_data(self):
...
-            except JSONDecodeError:
+                data = self.my_socket.recv(RECEIVE_SIZE)
                 if data == b'':
                     self.reconnect()
+                try:
+                    decoded_data = data.decode('utf-8')
+                except:
+                    continue
+                total_decoded += decoded_data
+                if sys.getsizeof(total_decoded) > MESSAGE_SIZE_BYTES:
```
Comment (AndreasBomholtz) [inline wappsto/connection/communication.py]: Why not just use 'len'?

File: wappsto/connection/communication.py [hunk 8]
```diff
@@ -623,23 +626,31 @@ def receive_data(self):
...
+                    error = "Received message exeeds size limit: {}".format(total_decoded)
```
Comment (AndreasBomholtz) [inline wappsto/connection/communication.py]: please do not print the large string when there is an error.

File: wappsto/connection/communication.py [hunk 9]
```diff
@@ -623,23 +626,31 @@ def receive_data(self):
...
-                total_decoded.append(decoded_data)
-            else:
-                break
-
-            try:
-                decoded = json.loads(''.join(total_decoded))
-            except JSONDecodeError:
+                data = self.my_socket.recv(RECEIVE_SIZE)
                 if data == b'':
                     self.reconnect()
```
Comment (AndreasBomholtz) [inline wappsto/connection/communication.py]: Should there not be a return here?

File: wappsto/connection/communication.py (hunk 7 above)
Comment (bubriks) [reply wappsto/connection/communication.py]: len(): returns us the number of elements present in the sequence
getsizeof(): check how much memory in bytes is allocated to store a complete object

the example is shown here: https://www.happyprogrammingguide.com/wp-content/uploads/2019/05/Python-check-length-and-size-of-string.-1024x576.jpg
